File: Sudoku_Cpp_Shell/src/Variable.cpp
#include"Variable.hpp"

/**
 * Represents a variable in a CSP
 */

int Variable::namingCounter = 1;

// =====================================================================
// Constructors
// =====================================================================

Variable::Variable ( Domain::ValueSet possible_Values, int row, int col, int block ) : domain(possible_Values)
{
	if ( domain.size() == 1 )
	{
		modified = true;
		changeable = false;
	}

	r = row;
	c = col;
	b = block;
	name = "v" + std::to_string(namingCounter++);
	oldSize = size();
	changeable = true;
}

Variable::Variable ( const Variable& v ) : domain(v.domain)
{
	r = v.row();
	c = v.col();
	b = v.block();
	modified = v.isModified();
	name = v.getName();
	changeable = v.isChangeable();
}

// =====================================================================
// Accessors
// =====================================================================

bool Variable::isChangeable( void ) const
{
	return changeable;
}

bool Variable::isAssigned( void ) const
{
	return size() == 1;
}

bool Variable::isModified( void ) const
{
	return modified;
}

int Variable::row ( void ) const
{
	return r;
}

int Variable::col ( void ) const
{
	return c;
}

int Variable::block ( void ) const
{
	return b;
}

int Variable::size ( void ) const
{
	return domain.size();
}

// Returns the assigned value or 0 if unassigned
int Variable::getAssignment ( void ) const
{
	if ( isAssigned() )
		return domain.getValues()[0];

	return 0;
}

Domain Variable::getDomain ( void ) const
{
	return domain;
}

std::string Variable::getName ( void ) const
{
	return name;
}

Domain::ValueSet Variable::getValues ( void ) const
{
	return domain.getValues();
}

bool Variable::operator== ( const Variable &other ) const
{
	return (row() == other.row()) && (col() == other.col()) && (block() == other.block());
}

bool Variable::operator!= ( const Variable &other ) const
{
	return !(*this == other);
}

// =====================================================================
// Modifiers
// =====================================================================

void Variable::setModified( bool modified )
{
	if ( ! changeable )
		return;

	this->modified = modified;
	this->domain.setModified(modified);
}

// Assign a value to the variable
void Variable::assignValue ( int val )
{
	if ( ! changeable )
		return;

	setDomain( Domain( val ) );
}

// Sets the domain of the variable
void Variable::setDomain ( Domain d )
{
	if ( ! changeable )
		return;

	domain = d;
	modified = true;
}

// Removes a value from the domain
void Variable::removeValueFromDomain ( int val )
{
	if ( ! changeable )
		return;

	domain.remove(val);
	modified = domain.isModified();
}

// =====================================================================
// Iterator
// =====================================================================

Domain::iterator Variable::begin()
{
	return domain.begin();
}

Domain::iterator Variable::end()
{
	return domain.end();
}

// =====================================================================
// String representation
// =====================================================================

std::string Variable::toString()
{
	std::stringstream ss;
	std::string sep = "";
	ss << " Name: " << name;
	ss << "\tdomain: {";

	for ( int i = 0; i < domain.getValues().size(); ++i )
	{
		ss << sep << domain.getValues()[i];
		sep = ",";
	}

	ss << "}";
	return ss.str();
}

File: Sudoku_Java_Shell/src/Variable.java
import java.util.Iterator;
import java.util.List;

/**
 * Represents a variable in a CSP
 */

public class Variable implements Iterable<Integer>
{
	// =================================================================
	// Properties
	// =================================================================

	private static int namingCounter = 1;
	private Domain domain;
	private int row, col, block;
	private boolean modified;
	private boolean changeable;
	private String name;

	private int oldSize;

	// =================================================================
	// Constructors
	// =================================================================

	public Variable ( List<Integer> possible_Values, int row, int col, int block )
	{
		this.domain = new Domain( possible_Values );
		if ( size() == 1 )
		{
			modified = true;
			changeable = false;
		}
		this.row = row;
		this.col = col;
		this.block = block;
		this.name = "v" + namingCounter++;
		this.oldSize = size();
		this.changeable = true;
	}

	public Variable ( Variable v )
	{
		this.domain = v.domain;
		this.row = v.row;
		this.col = v.col;
		this.block = v.block;
		this.modified = v.modified;
		this.name = v.name;
		this.changeable = v.changeable;
	}

	// =================================================================
	// Accessors
	// =================================================================

	public boolean isChangeable ( )
	{
		return changeable;
	}

	public boolean isAssigned ( )
	{
		return size() == 1;
	}

	public boolean isModified ( )
	{
		return modified;
	}

	public int row ( )
	{
		return row;
	}

	public int col ( )
	{
		return col;
	}

	public int block ( )
	{
		return block;
	}

	public int size ( )
	{
		return domain.size();
	}

	// Returns the assigned value or 0 if unassigned
	public Integer getAssignment ( )
	{
		if ( isAssigned() )
			return domain.getValues().get( 0 );

		return 0;
	}

	public Domain getDomain ( )
	{
		return domain;
	}

	public String getName ( )
	{
		return name;
	}

	public List<Integer> getValues ( )
	{
		return domain.getValues();
	}

	// =================================================================
	// Modifiers
	// =================================================================


	public void setModified ( boolean modified )
	{
		if ( ! changeable )
			return;

		this.modified = modified;
		this.domain.setModified( modified );
	}

	// Assign a value to the variable
	public void assignValue ( int val )
	{
		if ( ! changeable )
			return;

		setDomain( new Domain( val ) );
	}

	// Sets the domain of the variable
	public void setDomain ( Domain d )
	{
		if ( ! changeable )
			return;

		if ( ! domain.equals( d ) )
		{
			domain = d;
			modified = true;
		}
	}

	// Removes a value from the domain
	public void removeValueFromDomain ( int val )
	{
		if ( ! changeable )
			return;

		domain.remove( val );
		this.modified = domain.isModified();
	}

	// =================================================================
	// Iterator
	// =================================================================

	@Override
	public Iterator<Integer> iterator()
	{
		return domain.iterator();
	}

	// =================================================================
	// String Representation
	// =================================================================

	public String toString ( )
	{
		//prints node stats
		StringBuilder sb = new StringBuilder();
		sb.append(" Name: "+name);
		sb.append("\tdomain: {");

		for ( Integer i : domain )
			sb.append(i + ",");

		sb.deleteCharAt(sb.length()-1);
		sb.append("}");
		return sb.toString();
	}
}
